dao/src/main/java/se/fortnox/reactivewizard/db/transactions/Transaction.java

```diff
@@ -94,16 +99,25 @@ private void closeConnection(Connection connection) {
         }
     }
 
-    private void allFailed(Connection connection, Throwable throwable) {
-        for (TransactionStatement statement : statementsToExecute) {
+    private void allFailed(Throwable throwable) {
+        for (TransactionStatement transactionStatement : statementsToExecute) {
+            final Statement statement = transactionStatement.getAtomicStatement().get();
             try {
-                statement.getStatement().onError(throwable);
+                setTransactionToExecutable(transactionStatement, statement);
+                statement.onError(throwable);
             } catch (Exception onErrorException) {
                 log.error("onError threw exception", onErrorException);
             }
         }
     }
 
+    private void setTransactionToExecutable(TransactionStatement transactionStatement, Statement statement) {
+        final AtomicReference<Statement> atomicStatement = transactionStatement.getAtomicStatement();
+        final Transaction                transaction     = transactionStatement.getTransaction();
+        atomicStatement.compareAndSet(statement,null);
```

Comment:
If statement.onError would do this reset, then the imlplementation details would not have to bubble here and this whole method would not be needed.

Reply:
I do not have the necessary information in the statement on error method ( or I might miss something here). But the method has been removed and the code left is easier to read.